src/text.c

```diff
@@ -0,0 +1,394 @@
+	);
+}
+
+static bool text_save(Sdb *s, int fd, bool sort, SdbList *path) {
+	// path
+	write_path (fd, path);
+	write (fd, "\n", 1);
+
+	// k=v entries
+	SdbList *l = sdb_foreach_list (s, sort);
+	SdbKv *kv;
+	SdbListIter *it;
+	ls_foreach (l, it, kv) {
+		const char *k = sdbkv_key (kv);
+		const char *v = sdbkv_value (kv);
+
+		write_key (fd, k);
+		write (fd, "=", 1);
+		write_value (fd, v);
+
+		write (fd, "\n", 1);
+	}
+	ls_free (l);
+
+	// sub-namespaces
+	l = s->ns;
+	if (sort) {
+		l = ls_clone (l);
+		ls_sort (l, cmp_ns);
+	}
+	SdbNs *ns;
+	ls_foreach (l, it, ns) {
+		write (fd, "\n", 1);
+		ls_push (path, ns->name);
+		text_save (ns->sdb, fd, sort, path);
+		ls_pop (path);
+	}
+	if (l != s->ns) {
+		ls_free (l);
+	}
+
+	return true;
+}
+#undef FLUSH
+#undef ESCAPE_LOOP
+#undef ESCAPE
+
+SDB_API bool sdb_text_save_fd(Sdb *s, int fd, bool sort) {
+	SdbList *path = ls_new ();
+	if (!path) {
+		return false;
+	}
+	bool r = text_save (s, fd, sort, path);
+	ls_free (path);
+	return r;
+}
+
+SDB_API bool sdb_text_save(Sdb *s, const char *file, bool sort) {
+	int fd = open (file, O_WRONLY | O_CREAT | O_TRUNC | O_BINARY);
+	if (fd < 0) {
+		return false;
+	}
+	bool r = sdb_text_save_fd (s, fd, sort);
+	close (fd);
+	return r;
+}
+
+typedef enum {
+	STATE_NEWLINE,
+	STATE_PATH,
+	STATE_KEY,
+	STATE_VALUE
+} LoadState;
+
+typedef struct {
+	bool eof;
+	char *buf;
+	size_t bufsz;
+	Sdb *root_db;
+	Sdb *cur_db; // current namespace, changes when encountering a path line
+	size_t pos; // current processing position in the buffer
+	size_t line_begin;
+	size_t token_begin; // beginning of the currently processed token in the buffer
+	size_t shift; // amount to shift chars to the left (from unescaping)
+	SdbList/*<size_t>*/ *path;
+	LoadState state;
+	bool unescape; // whether the prev char was a backslash, i.e. the current one is escaped
+} LoadCtx;
+
+// to be called at the end of a line.
+// save all the data processed from the line into the database.
+static void load_flush_line(LoadCtx *ctx) {
+	ctx->unescape = false;
+	// finish up the line
+	ctx->buf[ctx->pos - ctx->shift] = '\0';
+	switch (ctx->state) {
+	case STATE_PATH: {
+		ls_push (ctx->path, (void *)ctx->token_begin);
+		SdbListIter *it;
+		void *token_off_tmp;
+		ctx->cur_db = ctx->root_db;
+		ls_foreach (ctx->path, it, token_off_tmp) {
+			size_t token_off = (size_t)token_off_tmp;
+			if (!ctx->buf[token_off]) {
+				continue;
+			}
+			ctx->cur_db = sdb_ns (ctx->cur_db, ctx->buf + token_off, 1);
+			if (!ctx->cur_db) {
+				ctx->cur_db = ctx->root_db;
+				break;
+			}
+		}
+		ls_destroy (ctx->path);
+		break;
+	}
+	case STATE_VALUE: {
+		const char *k = ctx->buf + ctx->line_begin;
+		const char *v = ctx->buf + ctx->token_begin;
+		if (!*k || !*v) {
+			break;
+		}
+		sdb_set (ctx->cur_db, k, v, 0);
+		break;
+	}
+	default:
+		break;
+	}
+	// prepare for next line
+	ctx->shift = 0;
+	ctx->state = STATE_NEWLINE;
+}
+
+static inline char unescape_raw_char (char c) {
```

Comment:
also unescape tabs. also add the reverse function and use it instead of the macros above

Reply:
Nope, there is no direct inverse.

Reply:
the function signature shouldnt hgave a space before (

Reply:
there's a inverse, but takes longer to explain than to implement, so not going to discuss further here. its easier for me to explain in another PR after merging this

Reply:
also this char c can be const char c, because it wont be modified inside. but thats just little picky details